### PR TITLE
Add full error messages in job meta

### PR DESCRIPTION
**Pull request by geophilusd**
Currently we only have the error message in the job details in morgue. Having the error backtrace also in the job details would make it easier to identify the cause for errors.

### CODE REVIEW

**Comment by darkleaf** [comment]
Having an error backtrace will significantly increase memory usage in Redis DB. 
I think you should use tools like [Sentry](https://github.com/getsentry/sentry-ruby) to inspect errors.

```ruby
Lowkiq.server_middlewares << -> (worker, batch, &block) do
  opts = {
    extra: {
      lowkiq: {
        worker: worker.name,
        batch: batch,
        thread_id: Thread.current.object_id
      }
    }
  }

  Raven.capture opts do
    block.call
  end
end
```

**Comment by ritikesh** [comment]
@darkleaf - can we leave that to the end user? the default state can be disabled(not store any backtraces) but users should be able to configure if they want to store backtraces. [Sidekiq does something similar](https://github.com/mperham/sidekiq/blob/615393d79402b00bec530ceb0e56c1b89527b7c0/lib/sidekiq/job_retry.rb#L48) as well. It helps to see a small backtrace of potential error causes within the dashboard for quick debugging. Also, not all users might have a monitoring tool integration by default during the initial stages. Apart from the redis memory usage, do you see any other downsides?

**Comment by darkleaf** [comment]
In BIA Lowkiq is high loaded and they have a lot of dead jobs. 
It's a real problem because they don't clear the dead queue and there isn't a dead job expiring mechanism.  

FYI, I haven't worked for BIA and I don't have a lot of time to develop and test new features.

Personally, I recommend you use monitoring tools. It's best practice.

About your contribution. I guess, it is required to update the UI too just because backtraces contain `\r\n`.

@solutus do you need backtraces in the UI?

**Comment by geophilusd** [comment]
@darkleaf The UI seems to be working fine with the backtraces.

As @ritikesh suggested, I introduced a setting `save_backtrace` to enable/disable storing messages along with backtraces so that users can configure this if needed.

**Comment by darkleaf** [comment]
> @darkleaf The UI seems to be working fine with the backtraces.

Can you attach a screenshot?
`full_message` has a complex structure:
```
irb(main):019:0> ex
=> #<RuntimeError: aaaaa aaa>
irb(main):020:0> ex.full_message
=> "(irb):15:in `<main>': \e[1maaaaa aaa (\e[1;4mRuntimeError\e[m\e[1m)\e[m\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb/workspace.rb:116:in `eval'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb/workspace.rb:116:in `evaluate'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb/context.rb:450:in `evaluate'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb.rb:567:in `block (2 levels) in eval_input'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb.rb:751:in `signal_status'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb.rb:548:in `block in eval_input'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb/ruby-lex.rb:251:in `block (2 levels) in each_top_level_statement'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb/ruby-lex.rb:233:in `loop'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb/ruby-lex.rb:233:in `block in each_top_level_statement'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb/ruby-lex.rb:232:in `catch'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb/ruby-lex.rb:232:in `each_top_level_statement'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb.rb:547:in `eval_input'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb.rb:481:in `block in run'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb.rb:480:in `catch'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb.rb:480:in `run'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/3.0.0/irb.rb:409:in `start'\n\tfrom /usr/local/Cellar/ruby/3.0.2_1/lib/ruby/gems/3.0.0/gems/irb-1.3.5/exe/irb:11:in `<top (required)>'\n\tfrom /usr/local/opt/ruby/bin/irb:23:in `load'\n\tfrom /usr/local/opt/ruby/bin/irb:23:in `<main>'\n"
```

This string will be passed in the following component as value argument:
```js
function Meta({label, value}) {
  return (
    <tr>
      <th>{label}</th>
      <td>{value}</td>
    </tr>
  );
}
```
There is no code to handle `\n` and `\e[1m` (colors?).

The second big problem is string size. Sidekiq use [compression](https://github.com/mperham/sidekiq/blob/615393d79402b00bec530ceb0e56c1b89527b7c0/lib/sidekiq/job_retry.rb#L255-L259).
And backtraces are [disabled by default](https://github.com/mperham/sidekiq/blob/615393d79402b00bec530ceb0e56c1b89527b7c0/lib/sidekiq/job_retry.rb#L48-L49).

> We don't store the backtrace by default as that can add a lot of overhead
> to the job and everyone is using an error service, right?

In conclusion, I advise you to use error services.

File: lib/lowkiq.rb [hunk 1]
```diff
@@ -109,4 +109,5 @@ def build_by_node_splitter(number_of_nodes, node_number)
   self.dump_payload = ::Marshal.method :dump
   self.load_payload = ::Marshal.method :load
   self.workers = []
+  self.save_backtrace = false
```
Comment (darkleaf) [inline lib/lowkiq.rb]: It will be better to use `self.format_error_message = :message.to_proc` and then `job[:error] = Lowkiq.format_error_message.call(ex)`
At least it requires no tests.
But the UI issues stay the same.

**Comment by solutus** [comment]
> @solutus do you need backtraces in the UI?

@darkleaf  it could be helpful in some applications.

**Comment by geophilusd** [comment]
@darkleaf made changes as per your suggestions.

Used this configuration for the below screenshot:
```
Lowkiq.format_error_message = Proc.new { |ex| ex.full_message(highlight: false) }
```
The new lines are properly handled. `highlight: false` removes the formatting which we didn't need here. Anyway, since this is now configurable, users can format it however they need.

Screenshot 2022-01-11 at 9 20 11 PM

Please review and let me know if I need to change anything. Also, should I update the documentation with more examples?

**Comment by darkleaf** [comment]
Ok. Let's complete the feature. `full_message(highlight: false)` seems to be fine.

1. We should have the ability to dump/load error messages. 

```ruby
Lowkiq.dump_error = ->(ex) { ex.message }
Lowkiq.load_error =  ->(ex) { ex }
```
It's will be awesome to add an example to the readme with a compressed implementation like it was implemented in Sidekiq (zip & base64). 
And please, do think about backward compatibility. Your new `load_error` fn should load both regular and compressed strings.

I guess it's required to change this [line](https://github.com/bia-technologies/lowkiq/blob/master/lib/lowkiq/queue/fetch.rb#L45):

```ruby
# error: x[1][2],
error: Lowkiq.load_error(x[1][2]),
```

Thus, we can specify the mechanism for storing errors.

2. We should display messages the right way.
I think there are two ways: `<pre>` and `dangerouslySetInnerHTML`. I like the first one.

```js
// frontend/src/dumb/details/job.js
// #48 // {isPresent(error)       && <Meta label="error"       value={error} />}
{isPresent(error)       && <Meta label="error"       value={<pre>error</pre>} />}
```

**Comment by geophilusd** [comment]
@darkleaf I have made some changes as per your suggestion. I added the methods to compress/uncompress in `lowkiq.rb`. 
```
3.0.2 :089 > error_message
 => "(irb):62:in `/': \e[1mdivided by 0 (\e[1;4mZeroDivisionError\e[m\e[1m)\e[m\n\tfrom (irb):62:in `test'\n\tfrom (irb):68:in `<main>'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb/workspace.rb:116:in `eval'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb/workspace.rb:116:in `evaluate'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb/context.rb:450:in `evaluate'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb.rb:567:in `block (2 levels) in eval_input'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb.rb:751:in `signal_status'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb.rb:548:in `block in eval_input'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb/ruby-lex.rb:251:in `block (2 levels) in each_top_level_statement'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb/ruby-lex.rb:233:in `loop'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb/ruby-lex.rb:233:in `block in each_top_level_statement'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb/ruby-lex.rb:232:in `catch'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb/ruby-lex.rb:232:in `each_top_level_statement'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb.rb:547:in `eval_input'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb.rb:481:in `block in run'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb.rb:480:in `catch'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb.rb:480:in `run'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/3.0.0/irb.rb:409:in `start'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/lib/ruby/gems/3.0.0/gems/irb-1.3.5/exe/irb:11:in `<top (required)>'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/bin/irb:23:in `load'\n\tfrom /Users/gdurairaj/.rvm/rubies/ruby-3.0.2/bin/irb:23:in `<main>'\n" 
3.0.2 :090 > compress_error(error_message).size
 => 509 
3.0.2 :091 > error_message.size
 => 1920 
3.0.2 :092 > uncompress_error(compress_error(error_message)) == error_message
 => true 
3.0.2 :093 > compressed?(compress_error(error_message))
 => true 
```
I checked the UI. It renders the error message correctly. Is there anything else that I need to test? Also can you give some pointers for adding test cases?

**Comment by darkleaf** [comment]
@geophilusd unfortunately, you didn't understand me. I've done the work. Please review and test it with your app.

**Comment by geophilusd** [comment]
@darkleaf I thought we would be compressing the error message by default for everyone. 
I tested it with our application and everything works fine.

**Comment by darkleaf** [comment]
Defaults will be the same.

**Comment by darkleaf** [comment]
@solutus 1.2.0 🎉

**Comment by darkleaf** [comment]
Thanks for your contribution!

**Comment by solutus** [comment]
@darkleaf , thanks!
